File: EIDSSAVR/eidss.core/Core/LookupCacheListener.cs
using System;
using System.Collections.Generic;
using System.Threading;
using bv.model.BLToolkit;
using bv.model.Model.Core;
using bv.common.Configuration;

namespace eidss.model.Core
{
    public static class LookupCacheListener
    {
        private static Timer m_Timer;
        private static long m_TimerInterval = 1000;

        public static void Listen()
        {
            if (m_Timer == null)
            {
                m_Timer = new Timer(TimerCallback, null, Timeout.Infinite, Timeout.Infinite);
            }
            m_Timer.Change(0, Timeout.Infinite);
        }
        public static void Stop()
        {
            if (m_Timer != null)
            {
                m_Timer.Change(Timeout.Infinite, Timeout.Infinite);
            }
        }
        public static void Cleanup()
        {
            GetModifiedLookupTables();
        }


        private static void TimerCallback(object state)
        {
            m_Timer.Change(Timeout.Infinite, Timeout.Infinite);
            try
            {
                var list = GetModifiedLookupTables();
                if (list != null)
                {
                    if (list.Count > 0)
                    {
                        foreach (var t in list)
                        {
                            LookupManager.ClearByTable(t);
                            LookupManager.ClearByTable("rft" + t);
                        }
                        LookupManager.ClearAndReloadOnIdle();
                    }
                }
            }
            catch (Exception)
            {
            }
            finally
            {
                m_Timer.Change(m_TimerInterval, Timeout.Infinite);
            }
        }

        private static List<string> GetModifiedLookupTables()
        {
            if (!BaseSettings.SuppressGettingModifiedLookupTables)
            {
                using (DbManagerProxy manager = DbManagerFactory.Factory.Create(ModelUserContext.Instance))
                {
                    try
                    {
                        return manager.SetSpCommand("dbo.spLookupTables_GetModified", ModelUserContext.ClientID).ExecuteScalarList<string>();
                    }
                    catch (Exception)
                    {
                    }
                }
            }
            return null;
        }
    }
}

File: EIDSSAVR/vb/Shared/bv_dbcommon/Core/LookupCacheListener.vb
Imports bv.common.Core
Imports bv.common.Objects
Imports System.Data
Imports System.Data.Common
Imports System.Collections.Generic
Imports System.Data.SqlClient
Imports System.Threading
Imports bv.model.Model.Core
Imports bv.common.Configuration

Namespace Core
    Public Class LookupCacheListener
        Private Shared m_Timer As System.Threading.Timer = Nothing
        Private Shared m_TimerInterval As Long = 1000
        Private Shared m_AsyncConnectionManager As ConnectionManager
        Public Shared Sub Listen()
            If m_Timer Is Nothing Then
                m_Timer = New System.Threading.Timer(AddressOf WorkerProc, Nothing, Timeout.Infinite, Timeout.Infinite)
            End If
            m_Timer.Change(0, Timeout.Infinite)
        End Sub
        Public Shared Sub [Stop]()
            'SyncLock m_Timer
            If Not m_Timer Is Nothing Then
                m_Timer.Change(Timeout.Infinite, Timeout.Infinite)
            End If
            'm_Timer.Dispose()
            'm_Timer = Nothing
            'End SyncLock
        End Sub
        Public Shared Sub CleanUp()
            GetModifiedLookupTables()
        End Sub
        Private Shared Sub WorkerProc(ByVal state As Object)
            If m_Timer Is Nothing Then
                Exit Sub
            End If
            'SyncLock m_Timer
            m_Timer.Change(Timeout.Infinite, Timeout.Infinite)
            Try
                Using dt As DataTable = GetModifiedLookupTables()
                    If Not dt Is Nothing Then
                        For Each row As DataRow In dt.Rows
                            LookupCache.Refresh(row("strInformationString").ToString)
                        Next
                    End If
                End Using
            Catch ex As Exception
                Dbg.Debug("error during lookup cache refreshing: {0}", ex.ToString)
            Finally
                If Not m_Timer Is Nothing Then
                    m_Timer.Change(m_TimerInterval, Timeout.Infinite)
                End If
            End Try
            'End SyncLock
        End Sub

        Private Shared Function GetModifiedLookupTables() As DataTable
            If BaseSettings.SuppressGettingModifiedLookupTables Or LookupCache.LookupTables.Count = 0 Then
                Return Nothing
            End If
            Dim cmd As IDbCommand = BaseDbService.CreateSPCommand("spLookupTables_GetModified", AsyncConnectionManager.Connection, Nothing)
            BaseDbService.AddParam(cmd, "@ClientID", ModelUserContext.ClientID)
            Dim dt As New DataTable
            SyncLock cmd.Connection
                Dim err As ErrorMessage = BaseDbService.FillTable(cmd, dt)
                If Not err Is Nothing Then
                    Dbg.Debug("error during receiving modified lookup tables: {0}", err.Exception.ToString)
                End If
                If dt.Rows.Count > 0 Then
                    Dbg.Debug("{0} lookup tables was modified", dt.Rows.Count)
                End If
            End SyncLock
            Return dt
        End Function

        Private Shared ReadOnly Property AsyncConnectionManager() As ConnectionManager
            Get
                If m_AsyncConnectionManager Is Nothing Then
                    m_AsyncConnectionManager = ConnectionManager.CreateNew(True)
                End If
                Return m_AsyncConnectionManager
            End Get
        End Property

    End Class
End Namespace
